custom/icds_reports/sqldata/exports/dashboard_usage.py

```diff
@@ -0,0 +1,177 @@
+         location
+        """
+        location_matrix = []
+        location_ids = []
+        for result in results_queryset:
+            row_data = [result['state_id'], result['district_id'], result['block_id'], result['supervisor_id'],
+                        result['doc_id']]
+            location_matrix.append(row_data)
+        # adding only descendants to the main location
+        sub_location_types = self.location_types[self.location_types.index(main_location_type) + 1:]
+        for sub_location in sub_location_types:
+            if result[sub_location] not in location_ids:
+                location_ids.append(result[sub_location])
+        return location_matrix, location_ids
+
+    def get_location_id_string_from_location_type(self, location_type):
+        if location_type == 'awc':
+            location_type = 'doc_id'
+        else:
+            location_type = location_type + '_id'
+        return location_type
+
+    def check_if_date_in_last_week(self, date):
+        if date is None:
+            return False
+        d = datetime.datetime.strptime(date, "%Y-%m-%d")
+        now = datetime.datetime.now()
+        return (d - now).days < 7
+
+    def is_launched_from_location_id(self, location_type, location_id):
+        filter_dict = {location_type + '_id': location_id, 'is_launched': 'yes'}
+        return AggAwc.objects.filter(**filter_dict).exists()
+
+    def get_excel_data(self):
+        excel_rows = []
+        filters = [['Generated at', india_now()]]
+        headers = ['Sr.No', 'State/UT Name', 'District Name', 'Block Name', 'Username', 'Level', 'Role',
+                   'Launched?', 'Last Login', 'Logged in the last week?', 'Total', 'Child', 'Pregnant Women',
+                   'Demographics', 'System Usage', 'AWC infrastructure', 'Child Growth Monitoring List',
+                   'ICDS - CAS Monthly Register', 'AWW Performance Report', 'LS Performance Report',
+                   'Take Home Ration']
+        excel_rows.append(headers)
+        serial_count = 0
+        for user_location in self.user.get_sql_locations(self.domain):
+            # getting the location types to retrieve for this user location
+            location_type_filter = {
+                self.get_location_id_string_from_location_type(user_location.location_type_name):
+                    user_location.get_id}
+            all_awc_locations = AwcLocation.objects.filter(**location_type_filter).values(*self.required_fields)
+            # converting the result set to matrix to fetch ancestors for a given location
+            location_matrix, location_ids =\
+                self.convert_rs_to_matrix(all_awc_locations, self.get_location_id_string_from_location_type(
+                    user_location.location_type_name))
+            users = self.get_users_by_location(location_ids)
+            records = list(ICDSAuditEntryRecord.objects.filter(url='/a/icds-dashboard-qa/cas_export')
```

Comment:
so i will bring only users under the logged in user location. or do i have to get the users in batches and add parallel processing?

Reply:
That is a good start, but is it not possible to do all the process once? Get a count for each user, then just look that number up in the user loop below. Instead of calculating that number each time.

Reply:
that is what happening there.The returned result set would be like this 
![Screenshot from 2019-10-21 18-37-37](https://user-images.githubusercontent.com/47347156/67207862-0327e880-f432-11e9-8ae6-d4056050b757.png). I am just looping to get the total count

Reply:
RIght, but you do that looping for each user in the report. What if you do that looping only once. Something like
```
user_counts = defaultdict(int)
for r in records:
    user_counts[r['username']] += 1
```
then use user_counts in the loop below

Reply:
ohh understood